### DISCUSSION TITLE
Graph compatibility between machines

### DISCUSSION THREAD

**Discussion by nilsnolde** [Project-OSRM/osrm-backend#6825] [Q&A]
Following https://github.com/Project-OSRM/osrm-backend/issues/6688#issuecomment-1697265053:

In which situations are OSRM graphs compatible with other machines (than they were produced on)? So far my understanding has been:
- no compat between commits/releases _necessarily_
- likely problems between architectures (i.e. x86 graph won't work on x64; or arm64 graph won't work on x64)
- compatibility of x64 across platforms, i.e. Linux x64 graphs run fine on Windows & Mac x64

Is the above more or less accurate? Or I can't expect Linux x64 graphs working on Windows x64? Or we don't really know and need to test?

**Comment by danpat** [comment]
I'm pretty sure that Linux<->Windows compatibility is 🚫 But it's been 20 years since I developed anything on Windows, so 🤷 

_Most_ systems these days are little endian, but different platforms have different ideas about the sizes of `int` and `long` types.  Until we purge the codebase of those and stick to explicit `std::uint32_t`/etc, code compiled on Linux and Windows may generate different sizes for `long`.

The "Data models" section on this page shows some of the combinations:

https://en.cppreference.com/w/cpp/language/types

If you move between platforms with the same C++ data models, then I think things will work OK.

We have explicit checks in the code to prevent use of files between versions - files built with 5.25 won't work with `osrm-routed` from 5.26.  This is by design, as we have no structure in the files to enable any easy forward/backward compatibility logic.

**Comment by nilsnolde** [comment]
Thanks @danpat. I'd be very interested to have cross-platform compatibility of the graph. Sounds like hell of a chore to convert the code base to use explicit types, but also a good way to get to know the code a bit better:)
